// @corbe30/fortune-excel/dist/ToFortuneSheet/FortuneBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FortuneImageBase = exports.FortunesheetCalcChain = exports.FortuneSheetConfigMerge = exports.FortuneSheetborderInfoCellValueStyle = exports.FortuneSheetborderInfoCellValue = exports.FortuneSheetborderInfoCellForImp = exports.FortuneConfig = exports.FortuneInlineString = exports.FortuneSheetCellFormat = exports.FortuneSheetCelldataValue = exports.FortuneSheetCelldataBase = exports.FortuneFileInfo = exports.FortuneSheetBase = exports.FortuneFileBase = void 0;
var FortuneFileBase = /** @class */ (function () {
    function FortuneFileBase() {
    }
    return FortuneFileBase;
}());
exports.FortuneFileBase = FortuneFileBase;
var FortuneSheetBase = /** @class */ (function () {
    function FortuneSheetBase() {
    }
    return FortuneSheetBase;
}());
exports.FortuneSheetBase = FortuneSheetBase;
var FortuneFileInfo = /** @class */ (function () {
    function FortuneFileInfo() {
    }
    return FortuneFileInfo;
}());
exports.FortuneFileInfo = FortuneFileInfo;
var FortuneSheetCelldataBase = /** @class */ (function () {
    function FortuneSheetCelldataBase() {
    }
    return FortuneSheetCelldataBase;
}());
exports.FortuneSheetCelldataBase = FortuneSheetCelldataBase;
var FortuneSheetCelldataValue = /** @class */ (function () {
    function FortuneSheetCelldataValue() {
    }
    return FortuneSheetCelldataValue;
}());
exports.FortuneSheetCelldataValue = FortuneSheetCelldataValue;
var FortuneSheetCellFormat = /** @class */ (function () {
    function FortuneSheetCellFormat() {
    }
    return FortuneSheetCellFormat;
}());
exports.FortuneSheetCellFormat = FortuneSheetCellFormat;
var FortuneInlineString = /** @class */ (function () {
    function FortuneInlineString() {
    }
    return FortuneInlineString;
}());
exports.FortuneInlineString = FortuneInlineString;
var FortuneConfig = /** @class */ (function () {
    function FortuneConfig() {
    }
    return FortuneConfig;
}());
exports.FortuneConfig = FortuneConfig;
var FortuneSheetborderInfoCellForImp = /** @class */ (function () {
    function FortuneSheetborderInfoCellForImp() {
    }
    return FortuneSheetborderInfoCellForImp;
}());
exports.FortuneSheetborderInfoCellForImp = FortuneSheetborderInfoCellForImp;
var FortuneSheetborderInfoCellValue = /** @class */ (function () {
    function FortuneSheetborderInfoCellValue() {
    }
    return FortuneSheetborderInfoCellValue;
}());
exports.FortuneSheetborderInfoCellValue = FortuneSheetborderInfoCellValue;
var FortuneSheetborderInfoCellValueStyle = /** @class */ (function () {
    function FortuneSheetborderInfoCellValueStyle() {
    }
    return FortuneSheetborderInfoCellValueStyle;
}());
exports.FortuneSheetborderInfoCellValueStyle = FortuneSheetborderInfoCellValueStyle;
var FortuneSheetConfigMerge = /** @class */ (function () {
    function FortuneSheetConfigMerge() {
    }
    return FortuneSheetConfigMerge;
}());
exports.FortuneSheetConfigMerge = FortuneSheetConfigMerge;
var FortunesheetCalcChain = /** @class */ (function () {
    function FortunesheetCalcChain() {
    }
    return FortunesheetCalcChain;
}());
exports.FortunesheetCalcChain = FortunesheetCalcChain;
var FortuneImageBase = /** @class */ (function () {
    function FortuneImageBase() {
    }
    return FortuneImageBase;
}());
exports.FortuneImageBase = FortuneImageBase;
